va/http.go

```diff
@@ -539,10 +539,19 @@ func (va *ValidationAuthorityImpl) processHTTPValidation(
 		// assign to the client transport in order to connect to the redirect target using
 		// the IP address we selected.
 		redirDialer, redirRecord, err := va.setupHTTPValidation(ctx, req.URL.String(), redirTarget)
-		records = append(records, redirRecord)
 		if err != nil {
 			return err
 		}
+
+		// Check for a redirect loop. If any URL is found twice before the
+		// redirect limit, return error.
+		for _, record := range records {
```

Comment:
This check should move higher in the function, around line 530. Specifically, I want to move it about `newHTTPValidationTarget`, which does some non-trivial work (in particular a DNS lookup).

Also, the `if` statement will then have to use `req.URL` instead of `record.URL`, which is a little clearer. Then the reader will know that it comes directly from the input `req` instead of having to wonder `setupHTTPValidation` modified or chose a different URL. Also, perhaps more to the point, it avoids treating `redirRecord`, which is intended as an output parameter, as an input/output parameter. Reducing the use of variables like that, which carry state through a function in surprising ways, makes code a lot clearer and easier to debug.

Reply:
Good catch! I have move the location of the check.